jxl/src/entropy_coding/ans.rs

```diff
@@ -2,6 +2,410 @@
+            ((1 << n) + bitstream.read(n as usize)?) as u8
+        } else {
+            0
+        })
+    }
+
+    fn read_prefix(br: &mut BitReader) -> Result<u16> {
+        Ok(match br.read(3)? {
+            0 => 10,
+            1 => {
+                for val in [4, 0, 11, 13] {
+                    if br.read(1)? != 0 {
+                        return Ok(val);
+                    }
+                }
+                12
+            }
+            2 => 7,
+            3 => {
+                if br.read(1)? != 0 {
+                    1
+                } else {
+                    3
+                }
+            }
+            4 => 6,
+            5 => 8,
+            6 => 9,
+            7 => {
+                if br.read(1)? != 0 {
+                    2
+                } else {
+                    5
+                }
+            }
+            _ => unreachable!(),
+        })
+    }
+}
+
+impl AnsHistogram {
+    #[inline(always)]
+    pub fn read(&self, br: &mut BitReader, state: &mut u32) -> Result<u32> {
+        let idx = *state & 0xfff;
+        let i = (idx >> self.log_bucket_size) as usize;
+        let pos = idx & self.bucket_mask;
+
+        let bucket = self.buckets[i];
+        let alias_symbol = bucket.alias_symbol as usize;
+        let alias_cutoff = bucket.alias_cutoff as u32;
+        let dist = bucket.dist as u32;
+
+        let map_to_alias = pos >= alias_cutoff;
+        let (offset, dist_xor) = if map_to_alias {
+            (bucket.alias_offset as u32, bucket.alias_dist_xor as u32)
+        } else {
+            (0, 0)
+        };
+
+        let dist = dist ^ dist_xor;
+        let symbol = if map_to_alias { alias_symbol } else { i };
+        let offset = offset + pos;
+
+        let next_state = (*state >> 12) * dist + offset;
+        let appended_state = (next_state << 16) | br.peek(16) as u32;
+        let select_appended = next_state < (1 << 16);
+        *state = if select_appended {
+            appended_state
+        } else {
+            next_state
+        };
+        br.consume(if select_appended { 16 } else { 0 })?;
+        Ok(symbol as u32)
+    }
+
+    // For optimizing fast-lossless case.
+    #[allow(unused)]
+    #[inline]
+    pub fn single_symbol(&self) -> Option<u32> {
+        self.single_symbol
+    }
+}
+
+#[derive(Debug)]
+pub struct AnsCodes {
+    histograms: Vec<AnsHistogram>,
+}
+
+impl AnsCodes {
+    pub fn decode(num: usize, log_alpha_size: usize, br: &mut BitReader) -> Result<AnsCodes> {
+        let histograms = (0..num)
+            .map(|_| AnsHistogram::decode(br, log_alpha_size))
+            .collect::<Result<_>>()?;
+        Ok(Self { histograms })
+    }
+
+    pub fn read(&self, br: &mut BitReader, ctx: usize, state: &mut u32) -> Result<u32> {
+        self.histograms[ctx].read(br, state)
+    }
+}
 
-//use crate::bit_reader::BitReader;
-//use crate::error::Error;
+#[inline]
+pub fn ans_validate(state: u32) -> Result<()> {
```

Comment:
It seems better to me to make an `AnsReader` (or similar) newtype around `u32` with a `read` function taking a `AnsCodes` and a `finalize` or similar method to check the final state, wdyt? This has the advantage of making the existence of a `AnsHistogram` type an implementation-detail, in case we ever want to avoid that i.e. for better performance.

Reply:
Newtype around ANS state sounds great, though I think `AnsHistogram` is already hidden implementation detail (it's a private type in `ans` module).